linien-gui: fix build

doronbehar opened the pull request:
~~Still a draft because a few more `packageOverrides` are needed.~~

## Description of changes


## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

doronbehar committed:
linien-gui: no with lib; in meta

doronbehar committed:
python312Packages.linien-client: no with lib; in meta

doronbehar committed:
python312Packages.linien-common: no with lib; in meta

doronbehar committed:
python312Packages.linien-common: explain how to use with numpy_1

doronbehar committed:
python312Packages.linien-client: use modern buildPythonPackage attributes

doronbehar committed:
python312Packages.linien-client: mark broken like linien-common

doronbehar committed:
linien-gui: fix build

dotlambda commented on pkgs/by-name/li/linien-gui/package.nix:
```diff
@@ -4,16 +4,44 @@
+      postPatch =
+        (old.postPatch or "")
+        # Extra empty line is here due to potential last line of existing
+        # postPatch ending with `\`.
```
Where is that the case? Should be fixed.

doronbehar replied:
> Where is that the case? Should be fixed.

In `pandas`. I was thinking about this at first, but then I realized it would cause > 1000 rebuilds...

dotlambda commented on pkgs/by-name/li/linien-gui/package.nix:
```diff
@@ -4,16 +4,44 @@
+    p: pattern:
+    p.overridePythonAttrs (old: {
+      postPatch =
+        (old.postPatch or "")
```
```suggestion
        old.postPatch or ""
```

doronbehar replied:
Are you sure that:

```nix
old.postPatch or "" + ''
  # more here
''
```

Won't be evaluated like:

```nix
old.postPatch or ("" + ''
  # more here
'')
```

?

dotlambda replied:
Not 100% sure. Is it?

doronbehar replied:
I'm not motivated enough to check, and this doesn't matter, as the `postPatch` hooks are modified not in a `packageOverrides` now.

dotlambda replied:
I just checked using `nix repl` and it woukd work without parentheses.

dotlambda commented on pkgs/by-name/li/linien-gui/package.nix:
```diff
@@ -4,16 +4,44 @@
+      #
+      # https://discourse.nixos.org/t/several-comments-about-priorities-and-new-policies-in-the-python-ecosystem/51790
+      pandas = loosenNumpy super.pandas "numpy>=2.0";
+      scipy = loosenNumpy super.scipy "numpy>=2.0.0,<2.5";
```
Doesn't it make sense to do this in the pandas and scipy packages? That way everybody can override numpy more easily.

dotlambda replied:
If you prefer we can do this in a separate PR. But I definitely think it should be done in those packages because the override might otherwise break with the next update.

doronbehar replied:
> If you prefer we can do this in a separate PR. But I definitely think it should be done in those packages because the override might otherwise break with the next update.

I prefer doing it all in this - single PR.

doronbehar committed:
python312Packages.pandas: remove leading `\` in postPatch

Makes adding a new command to `postPatch` not get interpreted as another
`substituteInPlace` argument.

doronbehar committed:
python312Packages.pandas: Allow building with Numpy 1

doronbehar committed:
python312Packages.scipy: Allow building with Numpy 1

doronbehar committed:
linien-gui: modernize arguments

dotlambda commented on pkgs/development/python-modules/pandas/default.nix:
```diff
@@ -84,10 +84,22 @@ let
+    # 2.x are supported. However upstream wants to always build with Numpy 2,
+    # and with it to still be able to run with a Numpy 1 or 2. We insist to
+    # perform this substitution even though python3.pkgs.numpy is of version 2
+    # nowadays, because our ecosystem unforetunatly doesn't allow easily
```
```suggestion
    # nowadays, because our ecosystem unfortunately doesn't allow easily
```

doronbehar replied:
Fixed typo in both comment blocks.

dotlambda commented on pkgs/development/python-modules/pandas/default.nix:
```diff
@@ -84,10 +84,22 @@ let
+    # that override globally the `numpy` attribute to point to `numpy_1`.
     postPatch = ''
       substituteInPlace pyproject.toml \
+        --replace-fail "numpy>2.0.0,<2.5" "numpy<2.5" \
```
Is the upper bound needed?
```suggestion
        --replace-fail "numpy>2.0.0,<2.5" numpy \
```

dotlambda commented on pkgs/development/python-modules/scipy/default.nix:
```diff
@@ -91,6 +91,21 @@ buildPythonPackage {
+  # that override globally the `numpy` attribute to point to `numpy_1`.
+  postPatch = ''
+    substituteInPlace pyproject.toml \
+      --replace-fail "numpy>2.0.0,<2.5" "numpy<2.5"
```
Same here
```suggestion
      --replace-fail "numpy>2.0.0,<2.5" numpy
```

doronbehar commented:
Thanks for that round of review :). I believe this should be ready now.

dotlambda commented on pkgs/by-name/li/linien-gui/package.nix:
```diff
@@ -4,16 +4,24 @@
 
-  nativeBuildInputs =
-    with python3.pkgs;
+  build-system =
```
Sorry, one more thing I overlooked. `wrapQtAppsHook` should be in `nativeBuildInputs`.

doronbehar replied:
> Sorry, one more thing I overlooked. `wrapQtAppsHook` should be in `nativeBuildInputs`.

You are correct! Thanks.

dotlambda reviewed (commented):
Feel free to merge yourself.

doronbehar commented:
Thanks!

FliegendeWurst commented on pkgs/development/python-modules/pandas/default.nix:
```diff
@@ -84,10 +84,22 @@ let
+    # that override globally the `numpy` attribute to point to `numpy_1`.
     postPatch = ''
       substituteInPlace pyproject.toml \
+        --replace-fail "numpy>2.0.0,<2.5" numpy \
```
Copy/paste error, proposed fix: #401523

FliegendeWurst commented on pkgs/development/python-modules/scipy/default.nix:
```diff
@@ -91,6 +91,21 @@ buildPythonPackage {
+  # that override globally the `numpy` attribute to point to `numpy_1`.
+  postPatch = ''
+    substituteInPlace pyproject.toml \
+      --replace-fail "numpy>2.0.0,<2.5" numpy
```
#402053

I am beginning to wonder "numpy>2.0.0,<2.5" came from in the first place ...

doronbehar replied:
> #402053
> 
> I am beginning to wonder where "numpy>2.0.0,<2.5" came from in the first place ...

From pandas :facepalm:
